packages/python/plotly/plotly/express/_core.py

```diff
@@ -454,7 +532,7 @@ def make_trace_kwargs(args, trace_spec, trace_data, mapping_labels, sizeref):
                             mapping = args["color_discrete_map"].copy()
                         else:
                             mapping = {}
-                        for cat in trace_data[attr_value]:
+                        for cat in trace_data.get_column(attr_value).to_list():
```

Comment:
For context, although `trace_data.get_column(attr_value)` is a Narwhals Series, which is an iterable, explicitly calling a `to_list()` makes sure that the elements we loop over are python objects in all cases.
Calling `to_py_scalar` for each element is significantly more time consuming for the cases in which elements are not python scalars (such as pyarrow)